### PR TITLE
Add course format features skels

### PR DESCRIPTION
**Pull request by ferranrecio**
Adding support for format plugin features and Moodle 4.0 versions.

### CODE REVIEW

**Comment by mudrd8mz** [comment]
Thank you @ferranrecio for preparing these patches. I left some comments above for the minor things I noticed. Please let me know if you would prefer to address them yourself, or if you are happy with me fixing them while merging this.

**Comment by mudrd8mz** [comment]
Sorry I did not realise the comments were in "Pending" state and I had to finish the review to publish them. Hopefully you can see them now.

File: classes/local/util/manager.php [hunk 1]
```diff
@@ -1145,6 +1149,66 @@ protected function prepare_contenttype_files() {
...
+     * Prepares the files for a course format plugin.
+     */
+    protected function prepare_format_files() {
+        $recipe = $this->recipe;
+        $formantname = str_replace('format_', '', $recipe['component_name'] ?? '');
+
+        // All formats below Moodle 4.0 are considered legacy formats as the way formats are build
+        // is not the same as in Moodle 3.x.
+        $requiredversion = $recipe['requires'] ?? '0';
+        if (intval(substr($requiredversion, 0, 1)) < 4) {
```
[the remark this answers is not in the document]
Comment (ferranrecio) [reply classes/local/util/manager.php]: Done.

File: skel/file/format/classes/output/courseformat/content.mustache [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    Contenttype content manager class.
+
+    Context variables required for this template:
+    * component
+    * copyright
+
+    Example context (json):
+    {
+        "component": "local_test",
+        "copyright": "2021 David Mudrák <david@moodle.com>"
```
[the remark this answers is not in the document]
Comment (ferranrecio) [reply skel/file/format/classes/output/courseformat/content.mustache]: I am not sure you don't deserve it, but I change it eitherway.

File: skel/file/format/classes/output/renderer.mustache [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    Context variables required for this template:
+    * component
+    * copyright
+
+    Example context (json):
+    {
+        "component": "local_test",
+        "copyright": "2021 David Mudrák <david@moodle.com>"
+    }
```
[the remark this answers is not in the document]
Comment (ferranrecio) [reply skel/file/format/classes/output/renderer.mustache:30]: Same here ;-)

File: classes/local/util/manager.php [hunk 4]
```diff
@@ -1145,6 +1149,66 @@ protected function prepare_contenttype_files() {
         }
     }
 
+    /**
+     * Prepares the files for a course format plugin.
+     */
+    protected function prepare_format_files() {
+        $recipe = $this->recipe;
+        $formantname = str_replace('format_', '', $recipe['component_name'] ?? '');
```
[the remark this answers is not in the document]
Comment (ferranrecio) [reply classes/local/util/manager.php]: Ups. Typo fixed.

**Comment by ferranrecio** [comment]
Hi @mudrd8mz ,

Do you have to oportunity to review the code? We are launching the new moodle devdocs page It will be great if course format plugins can be created using pluginskel (as documented there).

**Comment by mudrd8mz** [comment]
Yup, I'm on it. I'm just polishing up some existing code so that this new feature can be nicely tested and we get passing green CI prechecks. I shall do it tomorrow. Thank you for understanding and the patience.

**Comment by mudrd8mz** [comment]
p.s. the `assertRegExp()` related commit was dropped from your branch as there had been pending pull request #111 that addressed that, too. Please see the discussion there for why I'd been waiting with processing that one.

**Comment by mudrd8mz** [comment]
For the record, this is now part of the 1.6.0 version.

**Pull request closed without merging**